### PR TITLE
Backsert implementation (issue #456)

### PR DESCRIPTION
**Pull request by leandromoh**
just sharing an experimenting

### CODE REVIEW

**Comment by atifaziz** [comment]
I see you have `CountDown` embedded. 😆

**Comment by atifaziz** [comment]
📝 This is for #460.

**Comment by leandromoh** [comment]
@atifaziz I adjust the documentation. IMHO this PR is ready for merge. Some more in your opinion?

**Comment by leandromoh** [comment]
@atifaziz I adjusted the documentation, renamed operator to `Backsert` and updated README. 
Something before merge in your opinion?

**Comment by leandromoh** [comment]
> All tests should use a TestingSequence to make sure Backsert disposes and doesn't iterate the source sequence more than once.

done

File: MoreLinq/Backsert.cs [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+            if (second == null) throw new ArgumentNullException(nameof(second));
+            if (index < 0)  throw new ArgumentOutOfRangeException(nameof(index), "Index cannot be negative.");
+
+            return index == 0
+                   ? first.Concat(second)
+                   : _();
+
+            IEnumerable<T> _() =>
+                 first.CountDown(index, (e, cd) => (Element: e, Countdown: cd))
+                      .SelectMany((e, i) => i == 0
```
Comment (atifaziz) [inline MoreLinq/Backsert.cs]: This `SelectMany` use should be rolled-out by hand otherwise this version (which was fine as a prototype) will cause a heap allocation for _each_ item (`Enumerable.Repeat(e.Element, 1)`) of the first sequence!

Comment (leandromoh) [reply MoreLinq/Backsert.cs]: done!

Comment (atifaziz) [reply MoreLinq/Backsert.cs]: I made some improvements in b9643c3e517ca922fbcbb87a92754a7fab28e634 to simplify the tuple usage that were easier & faster to do than describe.

File: MoreLinq.Test/BacksertTest.cs [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+        public void Backsert(int count, int index)
+        {
+            var seq1 = Enumerable.Range(1, count);
+            var seq2 = new[] { 97, 98, 99 };
+
+            using (var test1 = seq1.AsTestingSequence())
+            using (var test2 = seq2.AsTestingSequence())
+            {
+                var result = test1.Backsert(test2, index);
+                var expectations = seq1.SkipLast(index).Concat(seq2).Concat(seq1.TakeLast(index));
```
Comment (atifaziz) [inline MoreLinq.Test/BacksertTest.cs]: I think this test (and others) rely too much on algorithmic expression of expectations that's hard to follow. It makes reviewing the test hard, that it's testing the right thing in the first place without doing the mental gymnastics. In other words, a literal expression is much clearer as to the inputs and what's going to happen:

```c#
[TestCase(new[] { 1, 2, 3 }, 1, new[] { 97, 98, 99 }, ExpectedResult = new[] { 1, 2, 97, 98, 99, 3 })]
[TestCase(new[] { 1, 2, 3 }, 2, new[] { 97, 98, 99 }, ExpectedResult = new[] { 1, 97, 98, 99, 2, 3 })]
public IEnumerable<int> Backsert(int[] seq1, int index, int[] seq2)
{
    using (var test1 = seq1.AsTestingSequence())
    using (var test2 = seq2.AsTestingSequence())
    {
        foreach (var item in test1.Backsert(test2, index))
            yield return item;
    }
}
```

Comment (leandromoh) [reply MoreLinq.Test/BacksertTest.cs]: I removed all "mental gymnastic"

**Comment by leandromoh** [comment]
@atifaziz is it ready for merge?

File: MoreLinq.Test/BacksertTest.cs (hunk 2 above)
Comment (atifaziz) [reply MoreLinq.Test/BacksertTest.cs]: > I removed all "mental gymnastic"

I'm afraid it's still there unless I'm missing something? For example:

```c#
var expectations = seq1.SkipLast(index).Concat(seq2).Concat(seq1.TakeLast(index));
```

Why not use `ExpectedResult` on `[TestCase]` instead? If you do, you can fold `BacksertWithIndexEqualsSourceLength` and `BacksertWithIndexZero` into the `Backset` test.

Comment (leandromoh) [reply MoreLinq.Test/BacksertTest.cs]: I know, but I prefer keep different tests for each scenario. I think explict tests improves clarity of what is being tested.

Comment (atifaziz) [reply MoreLinq.Test/BacksertTest.cs]: >  I think explict tests improves clarity of what is being tested.

It's not, unfortunately, which is why I am asking for the changes. I am still doing “mental gymnastics” to do the review so it's not helping. That means when I have 5 minutes and I try to make progress on a review for a _simple operator_ and 5 minutes become 15 and then I have to abort; that translates into the review staying open longer.

We have 3 tests with duplicate code that needs maintenance over time. The only explicit or unique part about each test is its name. If you wish to make a note of that, then I'd comment on top of the `[TestCase]` or even use its `Description` property.

Comment (leandromoh) [reply MoreLinq.Test/BacksertTest.cs]: I see. So basically "less is more". Okay. Adjust done.

File: MoreLinq.Test/BacksertTest.cs [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+        [TestCase(new[] { 1, 2, 3 }, 0, new[] { 8, 9 }, ExpectedResult = new[] { 1, 2, 3, 8, 9 })]
+        [TestCase(new[] { 1, 2, 3 }, 1, new[] { 8, 9 }, ExpectedResult = new[] { 1, 2, 8, 9, 3 })]
+        [TestCase(new[] { 1, 2, 3 }, 2, new[] { 8, 9 }, ExpectedResult = new[] { 1, 8, 9, 2, 3 })]
+        [TestCase(new[] { 1, 2, 3 }, 3, new[] { 8, 9 }, ExpectedResult = new[] { 8, 9, 1, 2, 3 })]
+        public IEnumerable<int> Backsert(int[] seq1, int index, int[] seq2)
+        {
+            using (var test1 = seq1.AsTestingSequence())
+            using (var test2 = seq2.AsTestingSequence())
+            {
+                foreach (var item in test1.Backsert(test2, index))
```
Comment (atifaziz) [inline MoreLinq.Test/BacksertTest.cs]: I think this loop can be replaced by simply a call to `ToArray`:

```c#
using (var test1 = seq1.AsTestingSequence())
using (var test2 = seq2.AsTestingSequence())
    test1.Backsert(test2, index).ToArray();
```

I am not sure there's anything to be gained by yielding the results in a streamed fashion.

Comment (leandromoh) [reply MoreLinq.Test/BacksertTest.cs]: done.
